### PR TITLE
Fix exclusivity for start offset in kinesis indexing service & check exclusivity properly in IndexerSQLMetadataStorageCoordinator

### LINKED ISSUE #7286: Duplicate data ingestion in Kinesis indexing service
### Affected Version

0.14.0-rc1

### Description

I created two dataSources for the same Kinesis stream, but it returned different results for the same query. Duplicate raw events was found in one dataSource (it was about 110 events). I'm suspecting [this line](https://github.com/apache/incubator-druid/blob/master/indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java#L2347). Maybe `isExclusive` should be set to true. But, I'm not sure why it happened for only one dataSource.

### PR DESCRIPTION
**Pull request by jihoonson**
Hopefully fixes #7286. The start offsets should be exclusive except for the very first task group.

Also, I added `SeekableStreamStartSequenceNumbers` and `SeekableStreamEndSequenceNumbers` to represent start and end offsets, respectively. Now, when publishing segments, the start offsets are checked against the last committed end offsets of the metadata store including the exclusivity. In metadata store, only `SeekableStreamEndSequenceNumbers` is stored to avoid storing unnecessary exclusivity information.

Since `SeekableStreamStartSequenceNumbers` and `SeekableStreamEndSequenceNumbers` are passed to tasks via ioConfig, `KafkaIndexTaskIOConfig` is to be changed to support them. It should be able to be deserialized from both the old version in the metadata store and the current version.
I think most of changes are made because of this, since I had to rename method names from `getStartPartitions`/`getEndPartitions` to `getStartSequenceNumbers`/`getEndSequenceNumbers`.

### CODE REVIEW

**Comment by jihoonson** [comment]
I found another bug of that `isExclusive` is missing in `SeekableStreamDataSourceMetadata`. I'll update this PR soon.

**Comment by jihoonson** [comment]
This PR is ready for review. I added `SeekableStreamStartSequenceNumbers` and `SeekableStreamEndSequenceNumbers` to represent start and end offsets, respectively. `SeekableStreamEndSequenceNumbers` is always stored in the metadata store. Before updating the metadata store, `SeekableStreamStartSequenceNumbers` of the new commit is checked against the last committed `SeekableStreamEndSequenceNumbers`.

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamEndSequenceNumbers.java [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+  public SeekableStreamEndSequenceNumbers(
+      final String stream,
+      final Map<PartitionIdType, SequenceOffsetType> partitionSequenceNumberMap
+  )
+  {
+    this(stream, null, partitionSequenceNumberMap, null);
+  }
+
+  public SeekableStreamStartSequenceNumbers<PartitionIdType, SequenceOffsetType> asStartPartitions(
```
Comment (clintropolis) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamEndSequenceNumbers.java]: Could you add a javadoc explaining the usage of this method?

Comment (jihoonson) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamEndSequenceNumbers.java]: Added, thanks.

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamStartSequenceNumbers.java [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+        (SeekableStreamStartSequenceNumbers<PartitionIdType, SequenceOffsetType>) other;
+
+    if (stream.equals(otherStart.stream)) {
+      // Same stream, merge sequences.
+      final Map<PartitionIdType, SequenceOffsetType> newMap = new HashMap<>(partitionSequenceNumberMap);
+      newMap.putAll(otherStart.partitionSequenceNumberMap);
+
+      // A partition is exclusive if it's
+      // 1) exclusive in "this" and it's not in "other"'s partitionSequenceNumberMap or
+      // 2) exslucive in "other"
```
[the remark this answers is not in the document]
Comment (jihoonson) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamStartSequenceNumbers.java]: Thanks, fixed.

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SequenceMetadata.java [hunk 3]
```diff
@@ -80,6 +83,7 @@ public SequenceMetadata(
     this.exclusiveStartPartitions = exclusiveStartPartitions == null
                                     ? Collections.emptySet()
                                     : exclusiveStartPartitions;
+    this.exclusiveEndOffset = exclusiveEndOffset == null;
```
[the remark this answers is not in the document]
Comment (jihoonson) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SequenceMetadata.java]: Good catch! This variable is not used anymore. I remove it. Thanks!

**Pull request merged**